cached_path/bytes_range.py

```diff
@@ -0,0 +1,140 @@
+
+if TYPE_CHECKING:
+    from rich.progress import Progress
+
+
+def get_bytes_range(
+    url_or_filename: PathOrStr,
+    index: int,
+    length: int,
+    cache_dir: Optional[PathOrStr] = None,
+    extract_archive: bool = False,
+    force_extract: bool = False,
+    quiet: bool = False,
+    progress: Optional["Progress"] = None,
+) -> bytes:
+    """
+    Get a range of up to ``length`` bytes starting at ``index``.
+
+    In some cases the entire file may need to be downloaded, such as when the server does not support
+    a range download or when you're trying to get a bytes range from a file within an archive.
+
+    .. caution::
+        You may get less than ``length`` bytes sometimes, such as when fetching a range from an HTTP
+        resource starting at 0 since headers will be omitted in the bytes returned.
+
+    Parameters
+    ----------
+
+    url_or_filename :
+        A URL or path to parse and possibly download.
+
+    index :
+        The index of the byte to start at.
+
+    length :
+        The number of bytes to read.
+
+    cache_dir :
+        The directory to cache downloads. If not specified, the global default cache directory
+        will be used (``~/.cache/cached_path``). This can be set to something else with
+        :func:`set_cache_dir()`.
+
+        This is only relevant when the bytes range cannot be obtained directly from the resource.
+
+    extract_archive :
+        Set this to ``True`` when you want to get a bytes range from a file within an archive.
+        In this case the ``url_or_filename`` must contain an "!" followed by the relative path of the file
+        within the archive, e.g. "s3://my-archive.tar.gz!my-file.txt".
+
+        Note that the entire archive has to be downloaded in this case.
+
+    force_extract :
+        If ``True`` and the resource is a file within an archive (when the path contains an "!" and
+        ``extract_archive=True``), it will be extracted regardless of whether or not the extracted
+        directory already exists.
+
+        .. caution::
+            Use this flag with caution! This can lead to race conditions if used
+            from multiple processes on the same file.
+
+    quiet :
+        If ``True``, progress displays won't be printed.
+
+        This is only relevant when the bytes range cannot be obtained directly from the resource.
+
+    progress :
+        A custom progress display to use. If not set and ``quiet=False``, a default display
+        from :func:`~cached_path.get_download_progress()` will be used.
+
+        This is only relevant when the bytes range cannot be obtained directly from the resource.
+    """
+    if not isinstance(url_or_filename, str):
+        url_or_filename = str(url_or_filename)
+
+    # If we're using the /a/b/foo.zip!c/d/file.txt syntax, handle it here.
+    exclamation_index = url_or_filename.find("!")
+    if extract_archive and exclamation_index >= 0:
```

Comment:
It's possible it could be part of the filename though, no? I think this is how `cached_path()` behaves so I wanted to match the behavior.

Reply:
What if an archive file has an `!` in its name? It's an edge case but would cause things to break. Having a separate optional parameter for the relative path of the file within the archive would be a way to get around the problem

Reply:
> Having a separate optional parameter for the relative path of the file within the archive would be a way to get around the problem

I agree but I don't want to deviate from the API of `cached_path()`. At least for now.